contracts/staking/contracts/src/staking_pools/MixinStakingPool.sol

```diff
@@ -122,36 +121,97 @@ contract MixinStakingPool is
+            );
+        }
+
         return poolId;
     }
 
+    function joinStakingPoolAsMaker(
+        bytes32 poolId
+    )
+        external
+    {
+        // Is the maker already in a pool?
+        address makerAddress = msg.sender;
+        if (isMakerAssignedToStakingPool(makerAddress)) {
+            LibRichErrors.rrevert(LibStakingRichErrors.MakerAddressAlreadyRegisteredError(
+                makerAddress
+            ));
+        }
+
+        IStructs.MakerPoolJoinStatus memory poolJoinStatus = IStructs.MakerPoolJoinStatus({
+            poolId: poolId,
+            confirmed: false
+        });
+        poolJoinedByMakerAddress[makerAddress] = poolJoinStatus;
+
+        // notify
+        emit PendingAddMakerToPool(
+            poolId,
+            makerAddress
+        );
+    }
+
     /// @dev Adds a maker to a staking pool. Note that this is only callable by the pool operator.
+    /// Note also that the maker must have previously called joinStakingPoolAsMaker.
     /// @param poolId Unique id of pool.
     /// @param makerAddress Address of maker.
-    /// @param makerSignature Signature proving that maker has agreed to join the pool.
     function addMakerToStakingPool(
         bytes32 poolId,
-        address makerAddress,
-        bytes calldata makerSignature
+        address makerAddress
     )
         external
         onlyStakingPoolOperator(poolId)
     {
-        // sanity check - did maker agree to join this pool?
-        if (!isValidMakerSignature(poolId, makerAddress, makerSignature)) {
-            LibRichErrors.rrevert(LibStakingRichErrors.InvalidMakerSignatureError(
-                poolId,
-                makerAddress,
-                makerSignature
-            ));
-        }
+        // Is the maker already in a pool?
         if (isMakerAssignedToStakingPool(makerAddress)) {
             LibRichErrors.rrevert(LibStakingRichErrors.MakerAddressAlreadyRegisteredError(
                 makerAddress
             ));
         }
 
-        poolIdByMakerAddress[makerAddress] = poolId;
+        // Is the maker trying to join this pool?
+        bytes32 makerPendingPoolId = poolJoinedByMakerAddress[makerAddress].poolId;
+        if (makerPendingPoolId != poolId) {
+            LibRichErrors.rrevert(LibStakingRichErrors.MakerNotPendingJoinError(
+                makerAddress,
+                makerPendingPoolId,
+                poolId
+            ));
+        }
+
+        // Is the pool already full?
+        if (getNumberOfMakersInStakingPool(poolId) == MAX_MAKERS_IN_POOL) {
+            LibRichErrors.rrevert(LibStakingRichErrors.PoolIsFullError(poolId));
+        }
+
+        // Add maker to pool
+        IStructs.MakerPoolJoinStatus memory poolJoinStatus = IStructs.MakerPoolJoinStatus({
+            poolId: poolId,
+            confirmed: true
+        });
+        poolJoinedByMakerAddress[makerAddress] = poolJoinStatus;
         makerAddressesByPoolId[poolId].push(makerAddress);
```

Comment:
What do we think about just getting rid of this state variable? `poolJoinedByMakerAddress` already tells us if a maker is part of a pool. I guess being able to query the addresses is helpful for the assertion wrappers, but is that it? You can always derive the list by examining event logs...

Reply:
Hm that's true... We use the length of the array to check the size of the pool, but we can do that with just a counter. @hysz thoughts?

Reply:
Yeah let's do it. Since the list is capped - if we did ever need this data on-chain - we could parameterize it and validate the makers in the same number of `sload` operations.